src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,257 @@
+		}
+	}
+}
+
+
+/*
+ * RecordSubplanExecutionsOnNodes iterates over the usedSubPlanNodeList,
+ * and for each entry, record the workerNodes that are accessed by
+ * the distributed plan.
+ *
+ * Later, we'll use this information while we broadcast the intermediate
+ * results to the worker nodes. The idea is that the intermediate result
+ * should only be broadcasted to the worker nodes that are accessed by
+ * the distributedPlan(s) that the subPlan is used in.
+ *
+ * Finally, the function recursively descends into the actual subplans
+ * of the input distributedPlan as well.
+ */
+void
+RecordSubplanExecutionsOnNodes(HTAB *intermediateResultsHash,
+							   DistributedPlan *distributedPlan)
+{
+	List *usedSubPlanNodeList = distributedPlan->usedSubPlanNodeList;
+	ListCell *usedSubPlanNodeCell = NULL;
+	List *subPlanList = distributedPlan->subPlanList;
+	ListCell *subPlanCell = NULL;
+	int workerNodeCount = GetWorkerNodeCount();
+
+	foreach(usedSubPlanNodeCell, usedSubPlanNodeList)
+	{
+		Const *resultIdConst = lfirst(usedSubPlanNodeCell);
+		Datum resultIdDatum = resultIdConst->constvalue;
+		char *resultId = TextDatumGetCString(resultIdDatum);
+
+		IntermediateResultsHashEntry *entry = SearchIntermediateResult(
+			intermediateResultsHash, resultId);
+
+		/* no need to traverse the whole plan if all the workers are hit */
+		if (list_length(entry->nodeIdList) == workerNodeCount)
+		{
+			elog(DEBUG4, "subplan %s is used in all workers", resultId);
+
+			break;
+		}
+		else
+		{
+			entry->nodeIdList = AppendAllAccessedWorkerNodes(entry->nodeIdList,
+															 distributedPlan,
+															 workerNodeCount);
+
+			elog(DEBUG4, "subplan %s is used in %lu", resultId, distributedPlan->planId);
+		}
+	}
+
+	/* descend into the subPlans */
+	foreach(subPlanCell, subPlanList)
+	{
+		DistributedSubPlan *subPlan = (DistributedSubPlan *) lfirst(subPlanCell);
+		CustomScan *customScan = FetchCitusCustomScanIfExists(subPlan->plan->planTree);
+		if (customScan)
+		{
+			DistributedPlan *distributedPlanOfSubPlan = GetDistributedPlan(customScan);
+			RecordSubplanExecutionsOnNodes(intermediateResultsHash,
+										   distributedPlanOfSubPlan);
+		}
+	}
+}
+
+
+/*
+ * AppendAllAccessedWorkerNodes iterates over all the tasks in a distributed plan
+ * to create the list of worker nodes that can be accessed when this plan is executed.
+ *
+ * If there are multiple placements of a Shard, all of them are considered.
+ */
+static List *
+AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedPlan, int
+							 workerNodeCount)
+{
+	List *taskList = distributedPlan->workerJob->taskList;
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = lfirst(taskCell);
+		ListCell *placementCell = NULL;
+		foreach(placementCell, task->taskPlacementList)
+		{
+			ShardPlacement *placement = lfirst(placementCell);
+			workerNodeList = list_append_unique_int(workerNodeList, placement->nodeId);
```

Comment:
ideally, we'd use a hash table for sets.

Reply:
We considered that as well, but couldn't make sure that it'd always (say more common case , worker count < 8) perform better.

Do you prefer that? If we had some more time, we wanted to benchmark this a bit, but couldn't get there :/